// spice.io/lib/plugins/email.postmark/index.js
var postmark = require('postmark'),
emailify     = require('emailify');

exports.plugin = function(router, params) {

	var mailer = postmark(params.apiKey);
	
	
	function sendEmail(req, res) {

		var ops = req.query;

		console.log('emailifying body');

		emailify.parse(ops.body, res.success(function(body) {


			console.log('sending email to "%s"', ops.to);
		
			mailer.send({
				To: ops.to,
				Subject: ops.subject,
				HtmlBody: body,
				From: ops.from || params.defaultEmail
			}, function(err, response) {

				try {

					var result = response;

					if(result.ErrorCode) {

						console.error(result.Message);
					} else {

						console.log(result.Message);
					}

				} catch(e) {

					console.error(e.stack)
				}
				
				res.end();
			});
		}));
			
		
	}


	function handleInbound(){};
	
	
	router.on({

		/**! 
		 * Sends an email off to postmark
		 * @param To where the email is going to
		 * @param Subject the subject of the email
		 * @param HtmlBody the body of the message
		 * @optional From where the email is coming from
		 */

		'pull send/email': sendEmail,

		/**
		 */

		'pull -method=POST postmark/inbound': handleInbound
	})
}
