// spice.io/lib/plugins/service.part.recursive/index.js
var net = require('psk/node/io/net'),   
Url    = require('url'),
loadRequest = net.loadRequest;

exports.plugin = function (router)
{
	function getSocialService (req, res)
	{           

		var loader = 
		{
			name:'SpiceKit',
			type:'graph',
			service: new (function()
			{      
				this.loadNewest = function(url,callback)
				{
					loadRequest(url,function(data)
					{
						try
						{
							callback(JSON.parse(data).result);
						}
						catch(e)
						{
							callback(null);
						}
					});
					//could be done more efficiently, but this is the easiest implementation so far. ALSO, different branded
					//versions of clove use different databases... which we CANNOT access directly from here. The database itself
					//could in the future be on an entirely different machine. 
				} 
			}
			)(),
			loaders:
			{
				default:
				{                                          
					group:
					{
						params: {url: {type: 'input'}},
						name: 'Subscription',
						ignoreStandardization: true,
						renderData: function (data)
						{            
							data.createdAt = new Date(data.createdAt);                
							return data;
						},   
						load:function(data, callback)
						{                        
							this.service.loadNewest(data.url, function (data)
							{                                   
								callback(!data,data);     
							})
						}
					}  
				}  
			}
		}

		res.end(loader)
	}
		
	router.on({
		'collect social/service': getSocialService
	})
}
