// spice.io/lib/plugins/service.part.instagram/index.js
var loadRequest = require('psk/node/io/net').loadRequest,
URLRequest = require('psk/node/io/net').Request;



exports.params = {clientId:{type:'input',name:'Client ID'},clientSecret:{type:'input',name:'Client Secret'}};

exports.plugin = function(router, params)
{

	var service = new (function()
	{				
		var clientId = params.clientId;
		var host = 'https://api.instagram.com/v1/';


		function load(uri,callback)
		{
			var request = new URLRequest(host+uri);
			request.data = {client_id:clientId};
			request.method = 'GET';
			// console.log(request.url)

			loadRequest(request,function(data)
			{
				try
				{
					callback(JSON.parse(data));
				}
				catch(e)
				{

				}
			}) 
		}

		this.searchTags = function(q,callback)
		{
			load('tags/'+q+'/media/recent',function(data)
			{
				callback(data ? data.data : null)
			})
		}


	}
	)()
	
	function getSocialService(req, res)
	{                
		var info = 
		{
			name:'Instagram',
			loaders: {
				photos:
				{
					search: 
					{
						name:'Photo Search'    
						,params:{q:{type:'input'}}
						,renderData:function(data)
						{
							var media;

							/*if(data.link)
							{
								var shortCode = data.link.match(/\/p\/(\w+)/)[1];
								media = [{
									type:'image'
									,link:'http://instagr.am/p/'+shortCode+'/media'
								}]
							}*/


							if(data.images)
							{
								media = media || [];
								var im = data.images.standard_resolution || data.images.low_resolution;
								data.link = im.url;
								media.push({type:'image',link:data.link||im.url,regular:{link:im.url,width:im.width,height:im.height}});
							}


							return {
								type:'image'
								,text:data.caption ? data.caption.text : ''
								,createdAt:new Date(Number(data.created_time)*1000)
								,link:data.link
								,likes:data.likes ? data.likes.count : 0
								,media:media
								,augment:false
							}	
						}   
						// ,ignoreMedia:true  
						,ignoreMedia:true
						,identify:
						{
							test:[/instagr.am\/#(\w+)/,
							/instagram.com\/#(\w+)/,
							/searchinstagram.com\/#(\w+)/,
							/instagram.heroku.com\/search?q=(\w+)/]
							,feed:function(match,callback)
							{
								callback(false, {q:match[1]});
							}
						}
						,findPromotions:function(item,callback)
						{
							item.promoting = [item.link];
							callback();
						}
						,load:function(data,callback)
						{
							service.searchTags(data.q,function(data)
							{
								callback(!data,data);
							})    
						}
					}

				}	
			}
		}


		info.loaders.default = info.loaders.photos;

		res.end(info)
	}
		
	function getNectars(req, res)
	{
		var nectars = [{     
			check:{
				domain:'instagram.com'
				,test:/p\/(\w+)\/?$/   
			}         
			,type:'image'
			,getMedia:function(matches,url)
			{                                            
				var id = matches[1]
				return {regular:{link:'http://instagr.am/p/'+id+'/media'}};
			}
		}];

		res.end(nectars);
	}
		
	router.on({
		'collect social/service': getSocialService,
		'collect nectar': getNectars
	})
}
