// spice.io/lib/plugins/service.part.flickr/index.js
var Flickr = require('./flickr').Flickr;


exports.plugin = function(router, params)
{
	
	var flickr = new Flickr('6babd262478a3261c776b1ffab8d4a4a','c8bd9623a9202323');
	
	function getSocialService(req, res)
	{

		// function rendwr

		function renderSearchData(rawData)
		{                      
			var md = {type:'image',regular:{link:rawData.url_s || rawData.url_m || rawData.url_l}};

			if(rawData.url_l) md.large  = {link:rawData.url_l};
			if(rawData.url_s) md.small  = {link:rawData.url_s};
			if(rawData.url_t) md.thumb  = {link:rawData.url_t};
			if(rawData.url_m) md.medium = {link:rawData.url_m};
			// if(rawData.url_s) md.square = {link:rawData.url_sq};

			var link = 'http://www.flickr.com/photos/'+rawData.owner+'/'+rawData.id;


			var photo = rawData.url_l || rawData.url_sq;   


			return {id:photo,
			type:'image',
			label:rawData.title, 
			createdAt:new Date(Number(rawData.dateupload)*1000), 
			icon:rawData.url_sq,
			link:link,
			augment:false,
			media:[md]};
		}


		var searchFlickr = 
		{
			name:'Search Photos'
			,params:{q:{type:'input'}}
			,renderData:renderSearchData
			,identify:
			[
			{ 
				test:[/flickr.com\/search\/.*q=([^&]+)/i]
				,feed:function(result,callback)
				{
					callback(false, {q:unescape(result[1])});
				}
			}
			]
			,load: function (data, callback)
			{
				this.service.searchPhoto(data.q, callback);          
			}
		}

		var info = 
		{
			type:'graph'
			,name:'Flickr'
			,service: new (function()
			{				
				var apiKey = params.apiKey;
				var flickr = new Flickr(params.apiKey, params.apiSecret);

				this.searchPhoto = function(keyword,callback)
				{
					flickr.searchPhoto({ text: escape(keyword), extras: 'date_upload,url_l,url_m,url_o,url_t,url_z,url_s,url_sq,owner_name' }, function (data)
					{                      
						var photos = [];

						if(data && data.photos)
						{
							photos = data.photos.photo;
						}                               

						callback(false, photos || []);
					})
				}
			}
			)(),
			loaders:
			{
				default:
				{
					search: searchFlickr
				}
				,photos:
				{
					search: searchFlickr
				}
			}
		}

		res.end(info);
	}
	
	
	function getNectars(req, res)
	{                           
		var nectars = 
		[
		{
			check:{
				domain:'flickr.com'  
				,test:/(\d+)\/?$/
			}
			,getMedia: function (matches, url) 
			{                       
				var id = matches[1];      
				var passThruUrl;
				
				console.log(url);
				
				mediator.pull('get.route.url', { command: 'flickrShowLarge', params: [id] }, function (url)
				{                   
					passThruUrl = url; 
				})                                              

				return { link: url, type: 'image', large: passThruUrl }
			} 
		}
		];

		res.end(nectars);
	}
	
	return;
	
	router.on({

		/**
		 */

		'collect social/service': getSocialService,

		/**
		 */
		 
		'collect route': getRoute,

		/**
		 */
		 
		'pull -http flickr/show/large/:id': function(request)
		{
			flickr.getPhotoUrl(request.data.id, function (url)
			{                                 
				if(!url) return request.end('not found');

				request.respond({ redirect: url });
			});
		}
		// getNectars: getNectars
	})
}
