// spice.io/lib/plugins/feeds.stream.core/db/models.js


module.exports = function(db)
{
    var Schema = db.base.Schema,
	ObjectId = Schema.ObjectId;
    
    var SpiceAppFeedInfo = new Schema({
	
        //id of the feed (service:category:action:params)
        _id		  	  : String
        
        //cache until this date, then unlock so we can load more
        , cacheUntil  : Date
        
        //last time this feed's been updated
        , lastUpdate  : Date
        
        //watch until this date, then stop loading
        , watchUntil  : Date
        
        //locks the feed from being loaded (means it's being loaded)
        , lockedUntil : Date
        
        //time to keep the feed alive
        , ttl   	  : Number
    });



    var SpiceAppPromotion = new Schema({
        
        _id:String, 
        
        link: { type: String, unique: true },
        
        site: String,
        
        //type of post
        type: String,
        
        //date published
        createdAt: Date,
        
        //service: facebook, twitter, etc.
        service: String, 
        
        //title, or label of post
        label: String,
        
        //message from user
        message: String,
        
        //text from article
        text: String,
        
        //number of times the items been checked for promos. change this
        lockCount: Number,
        
        //N likes
        likes: Number,
        
        //N dislikes
        dislikes: Number,
        
        //heat from likes vs publish date
        heat: Number,
        
        //more details on promotion / service
        promoDetails: {},
        
        feeds: {type: [String], index: -1 },
        
        //next time item's gonna be updated
        nextUpdate: Date,
        
        //the media for the array
        media: Array,
        
        //num comments to promotion
        numComments: Number, 
    });




	return {
		SpiceAppFeedInfo: db.model('feedInfo', SpiceAppFeedInfo),
		PromotedItemsModel: db.model('promotions', SpiceAppPromotion),
		db: db
	};
};
