// spice.io/lib/plugins/groups.part.saved.subscription.email/views.js

module.exports = function(fig) {

	var views = fig.views;

	views.NewsletterItemView = views.Template.extend({
	
		'tpl': '#newsletter-item-template',
		
		'override render': function()
		{
			this._super();
			
		},
		
		'templateData': function()
		{
			var article = this.ops.data,
			text = article.text.split(' '),
			media = article.media || article.metadata,
			coverImage;
			
			if(media)
			for(var i = 0, n = media.length; i < n; i++)
			{
				var mediaItem = media[i];
				
				if(mediaItem.type == 'image')
				{
					coverImage = mediaItem.link;
					if(mediaItem.primary) break;
				}
			}
			
			//max 60 words
			if(text.length > 60) text = text.slice(0, 60).concat('...');
			
			
			return {
				link: article.link,
				label: article.label,
				text: text.join(' '),
				coverImageUrl: coverImage
			}
		}
	})

	views.NewsletterView = views.Template.extend({
		
		/**
		 */
		
		'override render': function()
		{
			this._super();

			
			var primary = this.$$('#Primary')[0],
			self = this;

			
			this.ops.articles.forEach(function(article)
			{
				var articleContainer = self.document.createElement('div');
				
				var articleView = new self.router.views.NewsletterItemView({ el: articleContainer, data: article });
				
				self.addChild(articleView)
				
				primary.appendChild(articleContainer);
			})
		},
		
		
		
		'templateData': function()
		{
			return {
				title: this.ops.title,
				currentTime: this.ops.currentTime,
				curratedBy: this.ops.curatedBy,
				description: this.ops.description,
				unsubscribeUrl: this.ops.unsubscribeUrl,
				contact: this.ops.contact
			}
		}
	});

}
