// fig/lib/desktop_client/index.js
//#include ./plugins

var views      = require('./views'),
model          = require('./views/model'),
template       = require('./views/template'),
concrete       = require('./views/concrete')



/**
 * 
 */

var views = {
	View: concrete,
	Item: model.Item,
	Collection: model.Collection,
	Template: template.Template,
	ItemTemplate: template.ItemTemplate,
	CollectionTemplate: template.CollectionTemplate
}

/**
 */

exports.mixins = {
	errorHandling: require('./mixin/errorHandling')
}

/**
 * allow the views to be extended upon
 */

 if(typeof global != 'undefined') {
 	views = global.figViews = global.figViews || views;
 }

exports.views = views;

/**
 */
 

exports.plugin = function(router, params) {
	
	this.require(__dirname + "/plugins");


	
	var plviews = {},
	fig = {
		views: plviews
	};

	for(name in views) {
		var viewClazz = plviews[name] = views[name].extend({
			'override __construct': function() {
				this._super.apply(this, arguments);
				this.router = router;
				this.views = plviews;
			}
		});
	}


	router.on({
		
		'pull fig': function(req, res) {
			res.end(fig);
		}
	});

	router.push('fig', fig);

	router.views = plviews;


	return {
		views: fig
	}
}
