// spice.io/lib/plugins/comments.core/db/models.js

module.exports = function(db)
{
	var Schema = db.base.Schema,
	ObjectId = Schema.ObjectId;

	var CommentsInfo = new Schema({

		//the target comment, article, or saved article ID the comment's pointing to
		target: String,
		
		//type of comment: "comment", "article", "saved"
		type: String,
		
		//the user who made the post
		user: String,
		
		//additional user info
		userInfo: {},
		
		//the date the post was created at
		createdAt: Date,
		
		//number of people who liked the post
		likes: Number,
		
		//the number of people who disliked a post
		dislikes: Number,
		
		//difference between likes / dislikes
		score: Number,
		
		//the body of the comment
		text: String,
		
		//num replies to comment
		numComments: Number,
		
	});

	var CommentVoteInfo = new Schema({
		
		//the comment ID
		comment: String,
		
		//1=upvote, -1=downvote
		upvote: Boolean,
		
		//the user who voted the comment
		user: String,
	});

	return {
		CommentsModel: db.model('comments', CommentsInfo),
		CommentVotesModel: db.model('commentVotes', CommentVoteInfo)
	};
}

exports.db = function(db)
{
	
};
